redis/src/aio/pubsub.rs

```diff
@@ -0,0 +1,423 @@
+        }
+    }
+
+    /// Subscribes to a new channel.
+    pub async fn subscribe(&mut self, channel_name: impl ToRedisArgs) -> RedisResult<()> {
+        let mut cmd = cmd("SUBSCRIBE");
+        cmd.arg(channel_name);
+        self.send_recv(cmd.get_packed_command()).await
+    }
+
+    /// Unsubscribes from channel.
+    pub async fn unsubscribe(&mut self, channel_name: impl ToRedisArgs) -> RedisResult<()> {
+        let mut cmd = cmd("UNSUBSCRIBE");
+        cmd.arg(channel_name);
+        self.send_recv(cmd.get_packed_command()).await
+    }
+
+    /// Subscribes to a new channel with pattern.
+    pub async fn psubscribe(&mut self, channel_pattern: impl ToRedisArgs) -> RedisResult<()> {
+        let mut cmd = cmd("PSUBSCRIBE");
+        cmd.arg(channel_pattern);
+        self.send_recv(cmd.get_packed_command()).await
+    }
+
+    /// Unsubscribes from channel pattern.
+    pub async fn punsubscribe(&mut self, channel_pattern: impl ToRedisArgs) -> RedisResult<()> {
+        let mut cmd = cmd("PUNSUBSCRIBE");
+        cmd.arg(channel_pattern);
+        self.send_recv(cmd.get_packed_command()).await
+    }
+}
+
+/// A connection dedicated to pubsub messages.
+pub struct PubSub {
+    sink: PubSubSink,
+    receiver: tokio::sync::mpsc::UnboundedReceiver<Msg>,
+}
+
+async fn execute_connection_pipeline<T>(
+    codec: &mut T,
+    (pipeline, instructions): (crate::Pipeline, ConnectionSetupComponents),
+) -> RedisResult<AuthResult>
+where
+    T: Sink<Vec<u8>, Error = RedisError> + Stream<Item = RedisResult<Value>> + 'static,
+    T: Send + 'static,
+    T::Item: Send,
+    T::Error: Send,
+    T::Error: ::std::fmt::Debug,
+    T: Unpin,
+{
+    let count = pipeline.len();
+    if count == 0 {
+        return Ok(AuthResult::Succeeded);
+    }
+    codec.send(pipeline.get_packed_pipeline()).await?;
+
+    let mut results = Vec::with_capacity(count);
+    for _ in 0..count {
+        let value = codec.next().await;
+        match value {
+            Some(Ok(val)) => results.push(val),
+            _ => return Err(closed_connection_error()),
+        }
+    }
+
+    check_connection_setup(results, instructions)
+}
+
+async fn setup_connection<T>(
+    codec: &mut T,
+    connection_info: &RedisConnectionInfo,
+) -> RedisResult<()>
+where
+    T: Sink<Vec<u8>, Error = RedisError> + Stream<Item = RedisResult<Value>> + 'static,
+    T: Send + 'static,
+    T::Item: Send,
+    T::Error: Send,
+    T::Error: ::std::fmt::Debug,
+    T: Unpin,
+{
+    if execute_connection_pipeline(codec, connection_setup_pipeline(connection_info, true)).await?
+        == AuthResult::ShouldRetryWithoutUsername
+    {
+        execute_connection_pipeline(codec, connection_setup_pipeline(connection_info, false))
+            .await?;
+    }
+
+    Ok(())
+}
+
+impl PubSub {
+    /// Constructs a new `MultiplexedConnection` out of a `AsyncRead + AsyncWrite` object
+    /// and a `ConnectionInfo`
+    pub async fn new<C>(
+        connection_info: &ConnectionInfo,
```

Comment:
This should also be a RedisConnectionInfo as changes in https://github.com/redis-rs/redis-rs/pull/1270